assets/js/base/hooks/checkout/use-checkout-address.js

```diff
@@ -46,55 +43,54 @@ export const useCheckoutAddress = () => {
 			( ! customerId || isSameAddress( shippingAddress, billingData ) )
 	);
 
-	// Pushes to global state when changes are made locally.
-	useEffect( () => {
-		// Uses shipping address or billing fields depending on shippingAsBilling checkbox, but ensures
-		// billing only fields are also included.
-		const newBillingData = {
-			...( shippingAsBilling ? shippingAddress : billingFields ),
-			email: billingFields.email || billingData.email,
-			phone: billingFields.phone || billingData.phone,
-		};
-
-		if ( ! isEqual( newBillingData, billingData ) ) {
-			setBillingData( newBillingData );
-		}
-	}, [
-		billingFields,
-		shippingAsBilling,
-		billingData,
-		shippingAddress,
-		setBillingData,
-		setShippingAddress,
-	] );
+	const currentShippingAsBilling = useRef( shippingAsBilling );
+	const previousBillingData = useRef( billingData );
 
 	/**
-	 * Wrapper for updateBillingFields (from useState) which handles merging.
-	 *
-	 * @param {Object} newValues New values to store to state.
+	 * Sets shipping address data, and also billing if using the same address.
 	 */
-	const setBillingFields = useCallback(
-		( newValues ) =>
-			void updateBillingFields( ( prevState ) => ( {
-				...prevState,
-				...newValues,
-			} ) ),
-		[]
+	const setShippingFields = useCallback(
+		( value ) => {
+			setShippingAddress( value );
+
+			if ( shippingAsBilling ) {
+				setBillingData( value );
+			}
+		},
+		[ shippingAsBilling, setShippingAddress, setBillingData ]
 	);
 
-	const setEmail = ( value ) => void setBillingFields( { email: value } );
-	const setPhone = ( value ) => void setBillingFields( { phone: value } );
+	// This syncs billing data with shipping data if the checkbox is enabled.
+	useEffect( () => {
+		if ( currentShippingAsBilling.current !== shippingAsBilling ) {
+			if ( shippingAsBilling ) {
+				previousBillingData.current = billingData;
+				setBillingData( shippingAddress );
+			} else {
+				setBillingData( {
+					...previousBillingData.current,
+					email: undefined,
+					phone: undefined,
+				} );
```

Comment:
Won't this clear out previous set email and phone number if the `shippingAsBilling` checkbox is unchecked? Do we want that behaviour?

Reply:
I dont think so--we merge billing data, so existing emails are preserved. I'll double check tommorow.

Reply:
Confirmed this is a merge update.